dynamiqs/qarray.py

```diff
@@ -0,0 +1,558 @@
+from __future__ import annotations
+
+import warnings
+from abc import abstractmethod
+
+import equinox as eqx
+import jax.numpy as jnp
+import numpy as np
+import qutip as qt
+from jax import Array
+from jax.scipy.linalg import expm
+from jaxtyping import ArrayLike, ScalarLike
+
+from .utils.utils.general import (
+    cosm,
+    dag,
+    isbra,
+    isdm,
+    isket,
+    isop,
+    norm,
+    proj,
+    ptrace,
+    sinm,
+    tensor,
+    tobra,
+    todm,
+    toket,
+    unit,
+)
+
+__all__ = ['QArray', 'DenseQArray']
+
+
+# a decorator that takes a class method f and returns g(f(x))
+def pack_dims(method: callable) -> callable:
+    """Decorator to return a new QArray with the same dimensions as the original one."""
+
+    def wrapper(self: QArray, *args, **kwargs) -> callable:
+        return self.__class__(method(self, *args, **kwargs), dims=self.dims)
+
+    return wrapper
+
+
+class QArray(eqx.Module):
+    r"""Quantum array object. DenseQArray is a wrapper around JAX arrays. It offers
+    convenience methods to handle more easily quantum states
+    (bras, kets and density matrices) and quantum operators.
+    If you come from QuTiP, this is the equivalent of the Qobj class.
+    """
+
+    dims: tuple[int, ...]
+
+    @property
+    @abstractmethod
+    def dtype(self) -> jnp.dtype:
+        """Return the data type of the quantum state.
+
+        Returns:
+             The data type of the quantum state.
+        """
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, ...]:
+        """Returns the shape of the quantum state.
+
+        Returns:
+            The shape of the quantum state.
+        """
+
+    @property
+    def ndim(self) -> int:
+        """Returns the number of dimensions of the quantum state.
+
+        Returns:
+            The number of dimensions of the quantum state.
+        """
+        return len(self.dims)
+
+    @property
+    @abstractmethod
+    def I(self) -> QArray:  # noqa: E743
+        """Returns the identity operator compatible with the quantum state.
+
+        Returns:
+            The identity operator.
+        """
```

Comment:
I'm strongly against this. It understand that's it's a nice shortcut, but it makes no sense to me to read
```python
H = a @ a - a.I
# or
H = a @ a - psi0.I
# or whatever terrible things people can write with this
```
I think people should create the identity operator with `dq.eye()`.

Reply:
@gautierronan WDYT?

Reply:
I think I'm leaning towards the opinion of Pierre, but I'm not shocked either by this API.

IMO we could keep this API but never use it in the documentation (so it would only be documented in the class API). My reasoning is that some well-versed dynamiqsers might find it useful, but most users should indeed not use it because they will likely write terrible things with it.

Reply:
Let's discuss it live, I suggest removing it for now to move quickly on this PR. I don't think it's relevant to keep it for well-versed user, who share their code with beginners, who might misuse this.